### PR TITLE
Initial build for 1.0.0 ❄️

### PR DESCRIPTION
**Pull request by JeanChristopheMorinPerso**
Upstream: https://github.com/autogluon/autogluon/tree/v1.0.0

https://anaconda.atlassian.net/browse/PKG-3416

Channel: Snowflake

I suggest using https://inspector.pypi.io to get the metadata (dependencies) for each output:
* [autogluon](https://inspector.pypi.io/project/autogluon/1.0.0/packages/ce/7b/db15f5d33bb0bcdcaead1740ccb7018bc560a9fde14edd177c62f1051844/autogluon-1.0.0-py3-none-any.whl/autogluon-1.0.0.dist-info/METADATA)
* [autogluon.common](https://inspector.pypi.io/project/autogluon-common/1.0.0/packages/91/65/9d502252d8039755c32e40d04abe1dc019d10b4392f7981045ab14281d2f/autogluon.common-1.0.0-py3-none-any.whl/autogluon.common-1.0.0.dist-info/METADATA)
* [autogluon.core](https://inspector.pypi.io/project/autogluon-core/1.0.0/packages/d6/0b/d2d017b328a2101348c2a07752f57ae6f4c47bbb67e5527b10823f1db55b/autogluon.core-1.0.0-py3-none-any.whl/autogluon.core-1.0.0.dist-info/METADATA)
* [autogluon.features](https://inspector.pypi.io/project/autogluon-features/1.0.0/packages/a5/21/490e0a4310c7aaa22961c5885e2e03f0a6e8097d1d4af0f6c9b22d8fba50/autogluon.features-1.0.0-py3-none-any.whl/autogluon.features-1.0.0.dist-info/METADATA)
* [autogluon.tabular](https://inspector.pypi.io/project/autogluon-tabular/1.0.0/packages/98/30/8576a70043614aa1137a229916010b8080374fe131aeeee6323a93189b88/autogluon.tabular-1.0.0-py3-none-any.whl/autogluon.tabular-1.0.0.dist-info/METADATA)
* [autogluon.multimodal](https://inspector.pypi.io/project/autogluon-multimodal/1.0.0/packages/65/33/278e216cf910c9f7ef74ae9b4f2af4c405d254f41cef545485b508a2be6d/autogluon.multimodal-1.0.0-py3-none-any.whl/autogluon.multimodal-1.0.0.dist-info/METADATA)
* [autogluon.timeseries](https://inspector.pypi.io/project/autogluon-timeseries/1.0.0/packages/ec/6d/89b1b5b80f134753f98e335e57c6cb6d895add9887bf2ce16021b20bb6e2/autogluon.timeseries-1.0.0-py3-none-any.whl/autogluon.timeseries-1.0.0.dist-info/METADATA)

### CODE REVIEW

**Comment by Jrice1317** [comment]
Have you already tried implementing the linter suggestions? Not sure if this could be related to the issues you had with the patch and/or the tests failing.

File: recipe/meta.yaml [hunk 1]
```diff
@@ -1,178 +1,334 @@
...
-        - {{ pin_subpackage(name + '-core') }}
+        - python
+        - numpy <1.29,>=1.21
+        - scipy <1.13,>=1.5.4
+        - scikit-learn <1.5,>=1.3.0
+        - networkx <4,>=3.0
+        - pandas <2.2.0,>=2.0.0
+        - tqdm <5,>=4.38
+        - requests
+        - matplotlib
```
Comment (skupr-anaconda) [inline recipe/meta.yaml]:
```suggestion
        - matplotlib-base
```

File: recipe/meta.yaml [hunk 2]
```diff
@@ -1,178 +1,334 @@
...
+        - networkx <4,>=3.0
+        - {{ pin_subpackage(name + '.core', exact=True) }}
+        - {{ pin_subpackage(name + '.features', exact=True) }}
+        # Extra: all
+        - xgboost <2.1,>=1.6
+        - {{ pin_subpackage(name + '.core', exact=True) }}
+        - pytorch <2.1,>=2.0
+        - fastai <2.8,>=2.3.1
+        - lightgbm <4.2,>=3.3
+        - catboost <1.3,>=1.1
```
Comment (skupr-anaconda) [inline recipe/meta.yaml:272]: Is ray-core missing? See https://github.com/autogluon/autogluon/blob/caccb0b816e5e428370650f935da4e793aa23e12/tabular/setup.py#L59

File: recipe/meta.yaml (hunk 1 above)
Comment (skupr-anaconda) [reply recipe/meta.yaml]: autogluon.core fails for Python 3.11. Is there something incompatible?

Comment (JeanChristopheMorinPerso) [reply recipe/meta.yaml]: It needs ray, which is being built and is why it's failing.

**Comment by JeanChristopheMorinPerso** [comment]
@Jrice1317 The linter warnings are actually wrong. That's a linter bug.

**Comment by skupr-anaconda** [comment]
linux-64/autogluon.multimodal-1.0.0-py310h06a4308_0 failed because of unsatisfiable dependencies

File: recipe/meta.yaml [hunk 3]
```diff
@@ -1,178 +1,334 @@
...
     requirements:
       host:
+        - python
         - pip
-        - python <3.9
+        - setuptools
         - wheel
+      run:
+        - python
+        - {{ pin_subpackage(name + '.common', exact=True) }}
```
Comment (danpetry) [inline recipe/meta.yaml]: `~` in jinja is preferred for string concatenation, because it converts the operands into strings before concatenating them

File: recipe/meta.yaml [hunk 4]
```diff
@@ -1,178 +1,334 @@
...
+        - {{ pin_subpackage(name + '.core', exact=True) }}
+        - {{ pin_subpackage('autogluon.features', exact=True) }}
```
Comment (danpetry) [inline recipe/meta.yaml:44]: can you do `name ~` for these as well?

File: recipe/meta.yaml [hunk 5]
```diff
@@ -1,178 +1,334 @@
...
-        - matplotlib
-        - paramiko >=2.4
-        - dask >=2.6.0
-        - distributed >=2.6.0
-        - boto3
-        - autograd >=1.3
-        - dill >=0.3.3,<1.0
+        - python
+        - numpy <1.29,>=1.21
+        - pandas <2.2.0,>=2.0.0
```
Comment (danpetry) [inline recipe/meta.yaml:76]: just curious if there's any specific reason you're putting the range this way round?

File: recipe/meta.yaml [hunk 6]
```diff
@@ -1,178 +1,334 @@
...
-        - python <3.9
-        - wheel
         - setuptools
+        - wheel
       run:
-        - {{ pin_subpackage(name + '-core') }}
+        - python
+        - numpy <1.29,>=1.21
+        - scipy <1.13,>=1.5.4
+        - scikit-learn <1.5,>=1.3.0
```
Comment (danpetry) [inline recipe/meta.yaml:112]: you could put these dep ranges in jinja variables at the top maybe, then there'd be only one place you'd need to update them

File: recipe/meta.yaml [hunk 7]
```diff
@@ -1,178 +1,334 @@
...
+        - boto3 <2,>=1.10
+        - {{ pin_subpackage(name + '.common', exact=True) }}
+        # Extra: all (Required by autogluon.tabular)
+        # Requires-Dist: ray[default] <2.7,>=2.6.3 ; extra == 'all'
+        # Requires-Dist: ray[default,tune] <2.7,>=2.6.3 ; extra == 'all'
+        # Requires-Dist: async-timeout ; extra == 'all'
+        - async-timeout
+        # Requires-Dist: hyperopt <0.2.8,>=0.2.7 ; extra == 'all'
+
+        # This is autogluon.core[raytune]. Required by autogluon.multimodal.
```
Comment (danpetry) [inline recipe/meta.yaml]:
```suggestion
        # These are for autogluon.core[raytune]. Required by autogluon.multimodal.
```

File: recipe/meta.yaml [hunk 8]
```diff
@@ -1,178 +1,334 @@
...
+        - scipy <1.13,>=1.5.4
+        - scikit-learn <1.5,>=1.3.0
+        - networkx <4,>=3.0
+        - pandas <2.2.0,>=2.0.0
+        - tqdm <5,>=4.38
+        - requests
+        - matplotlib
+        - boto3 <2,>=1.10
+        - {{ pin_subpackage(name + '.common', exact=True) }}
+        # Extra: all (Required by autogluon.tabular)
```
Comment (danpetry) [inline recipe/meta.yaml]: I find these comments kind of difficult to understand. Could they be clarified/are they outdated?

File: recipe/meta.yaml [hunk 9]
```diff
@@ -1,178 +1,334 @@
...
+        - autogluon.core
       requires:
         - pip
         - pytest
-      imports:
-        - autogluon.features
       commands:
         - pip check
-        - cd features && pytest
+        - pytest -v core/tests -k 'not test_check_style'
```
Comment (danpetry) [inline recipe/meta.yaml]: Comments explaining the test omissions would be helpful, looks like a code style thing but could be nice to confirm

File: recipe/meta.yaml [hunk 10]
```diff
@@ -1,178 +1,334 @@
...
+        - {{ pin_subpackage(name + '.common', exact=True) }}
+        - pytorch-metric-learning <2.0,>=1.3.0
+        - nlpaug <1.2.0,>=1.1.10
+        - nltk <4.0.0,>=3.4.5
+        - openmim <0.4.0,>=0.3.7
+        - defusedxml <0.7.2,>=0.7.1
+        - jinja2 <3.2,>=3.0.3
+        - tensorboard <3,>=2.9
+        - pytesseract <0.3.11,>=0.3.9
+        # Exclusing because of imcompatibility with nvidia-ml-py.
```
Comment (danpetry) [inline recipe/meta.yaml]: I don't understand the comment - what's incompatible with nvidia-ml-py exactly?

File: recipe/meta.yaml [hunk 11]
```diff
@@ -1,178 +1,334 @@
...
+      source_files:
+        - multimodal/tests
+      imports:
+        - autogluon.multimodal
       requires:
         - pip
         - pytest
-        - openml
+      commands:
+        # TODO: Figure out why patch isn't working
```
Comment (danpetry) [inline recipe/meta.yaml]: just wondering if this is an outdated comment or still todo, and can we re-enable pip check

File: recipe/meta.yaml (hunk 2 above)
Comment (danpetry) [inline recipe/meta.yaml:272]: tabpfn, scikit-learn-intelex, imodels, vowpalwabbit, skil2onnx and onnxruntime seem to be missing too (considering the Extra: all comment)

File: recipe/patches/0001-Disable-mypy-pytest-plugin.patch [hunk 12]
```diff
@@ -0,0 +1,40 @@
+From 98e91e657750341971b9cf9a5394fd19b37c7a3f Mon Sep 17 00:00:00 2001
+From: Jean-Christophe Morin <jcmorin@anaconda.com>
+Date: Thu, 30 Nov 2023 13:39:57 -0500
+Subject: [PATCH] Disable mypy pytest plugin
```
Comment (danpetry) [inline recipe/patches/0001-Disable-mypy-pytest-plugin.patch]: a reason why would be helpful

**Review by danpetry — COMMENTED** [review]
I've thoroughly reviewed the dependencies.

There's a GPU version that can be built as well, I think it will be built if you include the gpu version of pytorch. (https://auto.gluon.ai/stable/install.html) However our pytorch-gpu v2.0.x doesn't currently support a major feature, so probably not ideal to pull it in, and I think we need torchvision GPU as well, so currently probably it wouldn't be possible anyway. Would it be possible to record a jira ticket for it, though? I can do it too if you want.

File: recipe/meta.yaml (hunk 5 above)
Comment (JeanChristopheMorinPerso) [reply recipe/meta.yaml:76]: I took all these from the wheels metadata. Considering the number of dependencies, I found that it reduces the mental overhead to have one to one what is in the wheels metadata.

File: recipe/meta.yaml (hunk 6 above)
Comment (JeanChristopheMorinPerso) [reply recipe/meta.yaml:112]: I feel like it would make the recipe harder to reason about than it already is.

File: recipe/meta.yaml (hunk 8 above)
Comment (JeanChristopheMorinPerso) [reply recipe/meta.yaml]: I'll clarify the comment. In the meantime, it means that it is the dependencies specified in the extra named "all".

File: recipe/meta.yaml (hunk 10 above)
Comment (JeanChristopheMorinPerso) [reply recipe/meta.yaml]: I'll clarify in the comment. `nvidia-ml-py` explicitly conflicts with `nvidia-ml` since they both provide the same files. `nvidia-ml` is deprecated and shouldn't be used.

File: recipe/meta.yaml (hunk 11 above)
Comment (JeanChristopheMorinPerso) [reply recipe/meta.yaml]: Outdated.

File: recipe/meta.yaml (hunk 2 above)
Comment (JeanChristopheMorinPerso) [reply recipe/meta.yaml:272]: `all` refers to the extra name, https://inspector.pypi.io/project/autogluon-tabular/1.0.0/packages/98/30/8576a70043614aa1137a229916010b8080374fe131aeeee6323a93189b88/autogluon.tabular-1.0.0-py3-none-any.whl/autogluon.tabular-1.0.0.dist-info/METADATA#line.44.

I could maybe add `run_constrained`? Though I gotta say that if we do that, this package is going to start to be a potential huge conflict nightmare...

File: recipe/meta.yaml (hunk 6 above)
Comment (danpetry) [reply recipe/meta.yaml:112]: ok, cool

File: recipe/meta.yaml (hunk 5 above)
Comment (danpetry) [reply recipe/meta.yaml:76]: ah ok, I was reading the github

File: recipe/meta.yaml (hunk 10 above)
Comment (danpetry) [reply recipe/meta.yaml]: where is nvidia-ml-py coming from?

File: recipe/meta.yaml (hunk 2 above)
Comment (danpetry) [reply recipe/meta.yaml:272]: no that's ok, ah I've been looking at the github while you've been looking at the wheel. Wheel's more accurate to what's really built but most people look at upstream code, just as an observation

File: recipe/meta.yaml (hunk 8 above)
Comment (danpetry) [reply recipe/meta.yaml]: see below, partly this is because I look at the github repo rather than the sdist contents

File: recipe/meta.yaml (hunk 10 above)
Comment (JeanChristopheMorinPerso) [reply recipe/meta.yaml]: It's coming from a transitive dep:
```
autogluon
  ray
    gpustat
      nvidia-ml-py
```

File: recipe/meta.yaml (hunk 8 above)
Comment (JeanChristopheMorinPerso) [reply recipe/meta.yaml]: Yep, I have local changes that will address all your comments, including this:
```
        # Extra group "all". autogluon.tabular requires autogluon.core[all]
        # Note that "all" depends on ray[default], ray[default,tune], async-timeout and hyperopt.
        - async-timeout

        # Extra group "raytune". autogluon.multimodal requires autogluon.core[raytune].
        - ray-tune <2.7,>=2.6.3
        - ray-default <2.7,>=2.6.3
        - hyperopt <0.2.8,>=0.2.7
```
which I hope is less cryptic.

File: recipe/meta.yaml (hunk 2 above)
Comment (danpetry) [reply recipe/meta.yaml:272]: would be nice to learn more about this kind of stuff if you have time sometime, maybe a workstream talk?

File: abs.yaml [hunk 13]
```diff
@@ -0,0 +1,30 @@
+upload_channels:
```
Comment (boldorider4) [inline abs.yaml:1]: All staging channels can be removed as they've already been merged (except for `pytorch-metric`)

File: recipe/meta.yaml (hunk 2 above)
Comment (JeanChristopheMorinPerso) [reply recipe/meta.yaml:272]: This is not ray-core per say. The `ray` extra requires `autogluon.core[all]`. Our `autogluon.core` package already depends on ray.

**Review by skupr-anaconda — APPROVED** [review]

**Review by danpetry — APPROVED** [review]

**Comment by danpetry** [comment]
well done 🎉
